File: solutions/src/main/java/com/adventofcode/aoc2022/Day2.java
package com.adventofcode.aoc2022;

import com.adventofcode.Solution;

import java.util.Map;
import java.util.stream.Stream;


public class Day2 extends Solution {

    @Override
    public Object part1(String input) {
        var matchScores = Map.of(
                "A X", 4, "A Y", 8, "A Z", 3,
                "B X", 1, "B Y", 5, "B Z", 9,
                "C X", 7, "C Y", 2, "C Z", 6
                );
        return Stream.of(input.split("\n"))
                .map(matchScores::get)
                .reduce(0, Integer::sum);
    }

    @Override
    public Object part2(String input) {
        var matchScores = Map.of(
                "A X", 3, "A Y", 4, "A Z", 8,
                "B X", 1, "B Y", 5, "B Z", 9,
                "C X", 2, "C Y", 6, "C Z", 7
        );
        return Stream.of(input.split("\n"))
                .map(matchScores::get)
                .reduce(0, Integer::sum);
    }
}

File: solutions/src/main/java/com/adventofcode/aoc2024/Day2.kt
package com.adventofcode.aoc2024

import com.adventofcode.Solution
import com.adventofcode.util.*
import kotlin.math.abs

class Day2 : Solution() {

    private fun safe(report: List<Int>) : Boolean {
        var lastValue : Int? = null
        var isIncreasing : Boolean? = null
        for (level in report) {
            if (lastValue == null) lastValue = level
            else {
                if (level == lastValue) return false
                if (isIncreasing == null) isIncreasing = level > lastValue
                if (isIncreasing && level < lastValue) return false
                if (!isIncreasing && level > lastValue) return false
                if (abs(level - lastValue) > 3) return false
                lastValue = level
            }
        }
        return true
    }

    override fun part1(input: String): Int {
        return input.nonEmptyLines()
            .map { report -> report.split(" ").map { it.toInt() } }
            .count { safe(it) }
    }

    override fun part2(input: String): Int {
        return input.nonEmptyLines()
            .map { report -> report.split(" ").map { it.toInt() } }
            .count { report ->
                val tolerantReports = sequence {
                    report.forEachIndexed { idx, _ ->
                        yield(report.filterIndexed{i, _ -> i != idx })
                    }
                }
                tolerantReports.any { safe(it) }
            }
    }
}